sdks/go/pkg/beam/io/rtrackers/offsetrange/offsetrange_test.go

```diff
@@ -0,0 +1,212 @@
+				}
+				// Check: All elements are still in a split restrictions. This
+				// logic assumes that the splits are returned in order which
+				// isn't guaranteed by EvenSplits, but this check is way easier
+				// with the assumption.
+				if split.Start != prevEnd {
+					t.Errorf("restriction range [%v, %v] missing after splits.",
+						prevEnd, split.Start)
+				} else {
+					prevEnd = split.End
+				}
+			}
+			if prevEnd != r.End {
+				t.Errorf("restriction range [%v, %v] missing after splits.",
+					prevEnd, r.End)
+			}
+		})
+	}
+}
+
+// TestTracker_TryClaim validates both success and failure cases for TryClaim.
+func TestTracker_TryClaim(t *testing.T) {
+	// Test that TryClaim works as expected when called correctly.
+	t.Run("Correctness", func(t *testing.T) {
+		tests := []struct {
+			rest   Restriction
+			claims []int64
+		}{
+			{rest: Restriction{Start: 0, End: 3}, claims: []int64{0, 1, 2, 3}},
+			{rest: Restriction{Start: 10, End: 40}, claims: []int64{15, 20, 50}},
+			{rest: Restriction{Start: 0, End: 3}, claims: []int64{4}},
+		}
+		for _, test := range tests {
+			test := test
+			t.Run(fmt.Sprintf("(rest[%v, %v], claims = %v)",
+				test.rest.Start, test.rest.End, test.claims), func(t *testing.T) {
+				rt := NewTracker(test.rest)
+				for _, pos := range test.claims {
+					ok := rt.TryClaim(pos)
+					// If TryClaim returns false, check if there was an error.
+					if !ok && !rt.IsDone() {
+						t.Fatalf("tracker claiming %v failed, error: %v", pos, rt.GetError())
+					}
+				}
+			})
+		}
+	})
+
+	// Test that each invalid error case actually results in an error.
+	t.Run("Errors", func(t *testing.T) {
+		tests := []struct {
+			rest   Restriction
+			claims []int64
+		}{
+			// Claiming backwards.
+			{rest: Restriction{Start: 0, End: 3}, claims: []int64{0, 2, 1}},
+			// Claiming before start of restriction.
+			{rest: Restriction{Start: 10, End: 40}, claims: []int64{8}},
+			// Claiming after tracker signalled to stop.
+			{rest: Restriction{Start: 0, End: 3}, claims: []int64{4, 5}},
+		}
+		for _, test := range tests {
+			test := test
+			t.Run(fmt.Sprintf("(rest[%v, %v], claims = %v)",
+				test.rest.Start, test.rest.End, test.claims), func(t *testing.T) {
+				rt := NewTracker(test.rest)
+				for _, pos := range test.claims {
+					ok := rt.TryClaim(pos)
+					// Finish successfully if we got an error.
+					if !ok && !rt.IsDone() && rt.GetError() != nil {
+						return
+					}
+				}
+				t.Fatal("tracker did not fail on invalid claim")
+			})
+		}
+	})
+}
+
+// TestTracker_TrySplit tests that TrySplit follows its contract, meaning that
+// splits don't lose any elements, split fractions are clamped to 0 or 1, and
+// that TrySplit always splits at the nearest integer greater than the given
+// fraction.
+func TestTracker_TrySplit(t *testing.T) {
+	tests := []struct {
+		rest     Restriction
+		claimed  int64
+		fraction float64
+		splitPt  int64 // Index where we want the split to happen.
```

Comment:
Is the splitPt here the "last element that should be processed" or "the first element that shouldn't be processed?"
Said another way, is the index represented by splitPt part of the Primary or part of the Residual?

Reply:
Reading into the test further, it looks like "last element of primary" which I'm pretty sure is the correct thing, so this just means the comment should be updated to reflect that.

Reply:
The range is actually half-closed range, so that split is the "first element that shouldn't be processed" of the primary, and the "first element to process" in the residual. But agreed, specified in the comment.